File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null){
            return head;
        }
        
        HashMap<Integer, Integer> map = new HashMap<>();
        ListNode curr = head;
        
        while (curr != null){
            int val = curr.val;
            map.put(val, map.getOrDefault(val, 0) + 1);
            curr = curr.next;
        }
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        curr = head;
        
        while (curr != null){
            int val = curr.val;
            if (map.containsKey(val) && map.get(val) > 1) {
                prev.next = curr.next;
            } else {
                prev = curr;
            }
            curr = curr.next;
        }
        return dummy.next;
    }
}

File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        seen = {}
        
        curr = head
        while curr:
            seen[curr.val] = seen.get(curr.val, 0) + 1
            curr = curr.next
        
        dum = ListNode(0)
        dum.next = head
        prev = dum
        curr = head
        
        while curr:
            if seen[curr.val] > 1:
                while curr.next and curr.next.val == curr.val:
                    curr = curr.next
                prev.next = curr.next
                # prev = prev.next
                # curr = curr.next
                curr = curr.next
            else:
                prev = curr
                curr = curr.next
        
        return dum.next